src/operations/extension/reduction_partial_rows.hpp

```diff
@@ -0,0 +1,237 @@
+                                 get_num_workgroup_cluster(compute_units));
+    const cl::sycl::range<1> wgs(work_group_size);
+    return cl::sycl::nd_range<1>(nwg * wgs, wgs);
+  }
+
+  template <typename local_memory_t>
+  SYCL_BLAS_INLINE void eval(local_memory_t scratch,
+                             cl::sycl::nd_item<1> id) noexcept {
+    /* reference to the scratch memory */
+    auto scratch_ptr = scratch.localAcc.get_pointer().get();
+
+    /* workgroup id */
+    const index_t group_id = id.get_group(0);
+    /* Local thread id */
+    const index_t local_id = id.get_local_id(0);
+
+    /* Block row and column */
+    const index_t group_col = group_id / group_count_rows_;
+    const index_t group_row = group_id - group_col * group_count_rows_;
+
+    /* Item row and column within a block */
+    const index_t local_col = local_id / work_group_rows;
+    const index_t local_row = local_id - local_col * work_group_rows;
+
+    /* Global position of the first element processed by the thread */
+    const index_t global_row = group_row * local_memory_rows + local_row;
+    const index_t global_col = group_col * local_memory_cols + local_col;
+
+    /* Total number of item cols in all work groups */
+    const index_t total_item_cols = local_memory_cols * group_count_cols_;
+
+    /* Initialize private reduction registers */
+    element_t accumulators[rows_per_item];
+#pragma unroll
+    for (index_t wpr = 0; wpr < rows_per_item; wpr++) {
+      accumulators[wpr] = init_val;
+    }
+
+    /* Sequential reduction level:
+     * Load multiple elements from the global memory, reduce them together and
+     * store them in the local memory */
+    {
+      index_t elem_col = global_col;
+      while (elem_col < cols_) {
+        index_t elem_col_idx = leading_dim_ * elem_col;
+        /* Each thread is responsible for multiple independent rows */
+        index_t elem_row = global_row;
+#pragma unroll
+        for (index_t wpr = 0; wpr < rows_per_item; wpr++) {
+          if (elem_row < rows_) {
+            const value_t lhs_val = accumulators[wpr];
+            const value_t rhs_val =
+                in_.template eval<true>(elem_col_idx + elem_row);
+            accumulators[wpr] = operator_t::eval(lhs_val, rhs_val);
+          }
+          elem_row += work_group_rows;
+        }
+        elem_col += total_item_cols;
+      };
+    }
+
+    /* Copy the accumulation registers into the local memory */
+    {
+      index_t local_memory_idx = local_memory_rows * local_col + local_row;
+#pragma unroll
+      for (index_t wpr = 0; wpr < rows_per_item; wpr++) {
+        scratch_ptr[local_memory_idx] = accumulators[wpr];
+        local_memory_idx += work_group_rows;
+      }
+    }
+
+    /* Parallel-reduction level:
+     * Tree-based reduction in local memory */
+    {
+      const index_t local_memory_lhs = local_col * local_memory_rows;
+#pragma unroll
+      for (index_t stride = work_group_cols / 2; stride > 0; stride /= 2) {
```

Comment:
similarly because you are using the tile this means that you need to have 2 nested for loop here. on devices which has enough registers, this might give better performance as you unroll the for loop and creating more tasks per thread, however, in devices with limited number of register, using too much register may cause register spill and can cause performance drop. If you wanted to use this technique at least provide a macro  or  compile time template parameters to disable it for devices with small number of registers

Reply:
Solved